yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+    return { preimage: O.fromBuffer(buffer), index };
+  }
+
+  /**
+   * This is wrapper around treeId to set the correct minimum leaf preimage
+   */
+  private setMinInfo<ID extends IndexedTreeId>(treeId: ID, preimage: { toBuffer: () => Buffer }, index: bigint): void {
+    this.indexedTreeMin.set(treeId, [preimage.toBuffer(), index]);
+  }
+
+  /**
+   * This is wrapper around treeId to set values in the indexedUpdates map
+   */
+  private setIndexedUpdates<ID extends IndexedTreeId>(treeId: ID, index: bigint, O: { toBuffer: () => Buffer }): void {
+    let updates = this.indexedUpdates.get(treeId);
+    if (updates === undefined) {
+      updates = new Map();
+      this.indexedUpdates.set(treeId, updates);
+    }
+    updates.set(index, O.toBuffer());
+  }
+
+  /**
+   * This is wrapper around treeId to get values in the indexedUpdates map
+   */
+  private getIndexedUpdates<ID extends IndexedTreeId, O>(
+    treeId: ID,
+    index: bigint,
+    O: { fromBuffer: (buffer: Buffer) => O },
+  ): O {
+    const updates = this.indexedUpdates.get(treeId);
+    if (updates === undefined) {
+      throw new Error('No updates found');
+    }
+    const buffer = updates.get(index);
+    if (buffer === undefined) {
+      throw new Error('No updates found');
+    }
+    return O.fromBuffer(buffer);
+  }
+
+  /**
+   * This is wrapper around treeId to check membership (i.e. has()) of index in the indexedUpdates map
+   */
+  private hasLocalUpdates<ID extends IndexedTreeId>(treeId: ID, index: bigint): boolean {
+    const updates = this.indexedUpdates.get(treeId);
+    if (updates === undefined) {return false;}
+    return updates.has(index);
+  }
+
+  /**
+   * This gets the low leaf preimage and the index of the low leaf in the indexed tree given a value (slot or nullifier value)
+   * If the value is not found in the tree, it does an external lookup to the merkleDB
+   * @param treeId - The tree we are looking up in
+   * @param value - The value for which we are look up the low leaf for.
+   * @param T - The type of the preimage (PublicData or Nullifier)
+   * @returns The low leaf preimage and the index of the low leaf in the indexed tree
+   */
+  async getLowInfo<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    value: Fr,
+    T: { fromBuffer: (buffer: Buffer) => T },
+  ): Promise<LowPreimageWitness<T>> {
+    // This can probably be done better, we want to say if the minInfo is undefined (because this is our first operation) we do the external lookup
+    const minPreimage = this.getMinInfo(treeId, T);
+    const start = minPreimage?.preimage;
+    // If the first element we have is already greater than the value, we need to do an external lookup
+    if (minPreimage === undefined || (start?.getKey() ?? 0n) >= value.toBigInt()) {
+      // The low public data witness is in the previous tree
+      const { index: lowPublicDataIndex } = (await this.treeDb.getPreviousValueIndex(treeId, value.toBigInt()))!;
```

Comment:
I thought that this returns the low leaf only if the target key (leafSlot/nullifier) doesn't exist?